File: huipyo/src/week05/boj2579/Main.java
package week05.boj2579;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int count = scanner.nextInt();
        int[] array = initArray(count, scanner);
        LinkedList<Integer> stack = new LinkedList<>();
        AtomicInteger answer = new AtomicInteger(0);
        Runnable backtracking = new Runnable() {
            @Override
            public void run() {
                for (int i : new int[] {1, 2}) {
                    if (i == 1 && !stack.isEmpty() && stack.peek() == 1) continue;
                    stack.push(i);
                    if (isSumEqualsCount(stack, count)) {
                        var sum = array[0];
                        var jSum = 0;

                        for (int j : stack) {
                            jSum += j;
                            sum += array[jSum];
                            answer.set(Math.max(answer.get(), sum));
                        }
                    }
                    if (stack.size() < count) {
                        run();
                    }
                    stack.pop();
                }
            }
        };

        backtracking.run();
        System.out.println(answer.get());
        scanner.close();
    }

    private static int sum(List<Integer> list) {
        int result = 0;

        for (int e : list) {
            result += e;
        }
        return result;
    }

    private static boolean isSumEqualsCount(List<Integer> list, int count) {
        int sum = sum(list);
        return sum == count - 1 || sum == count;
    }

    private static int[] initArray(int count, Scanner scanner) {
        int[] array = new int[count + 1];

        for (int i = array.length - 1; i > 0; i--) {
            array[i] = scanner.nextInt();
        }
        return array;
    }
}

File: huipyo/src/week05/boj2579/Main.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
let input = fs.readFileSync(filePath).toString().trim().split('\n');
let count = parseInt(input.shift());
var array = initArray(count);
var stack = [];
var answer = 0;

function backtracking() {
    for (let i of [1, 2]) {
        if (i === 1 && stack.length && stack[stack.length - 1] === 1) continue;
        stack.push(i);
        if (isSumEqualsCount(stack, count)) {
            let sum = array[0];
            let jSum = 0;

            for (let j of stack) {
                jSum += j;
                sum += array[jSum];
                answer = Math.max(answer, sum);
            }
        }
        if (stack.length < count) {
            backtracking();
        }
        stack.pop();
    }
}

backtracking();
console.log(answer);

function sum(arr) {
    return arr.reduce((acc, curr) => acc + curr, 0);
}

function isSumEqualsCount(arr, count) {
    var s = sum(arr);
    return s === count - 1 || s === count;
}

function initArray(count) {
    var arr = new Array(count + 1);

    for (let i = arr.length - 1; i > 0; i--) {
        arr[i] = parseInt(input.shift());
    }
    return arr;
}